bokeh/document.py

```diff
@@ -531,6 +532,100 @@ def _initialize_references_json(cls, references_json, references):
 
             instance.update_from_json(obj_attrs, models=references)
 
+    @classmethod
+    def _event_for_attribute_change(cls, changed_obj, key, new_value, value_refs, doc):
+      event = dict(
+        kind='ModelChanged',
+        model=dict(id=changed_obj['id'], type=changed_obj['type']),
+        attr=key,
+        new=new_value,
+      )
+      HasProps._json_record_references(doc, new_value, value_refs, True) # True = recurse
+      return event
+
+    @classmethod
+    def _events_to_sync_objects(cls, from_obj, to_obj, value_refs, doc):
+        from_keys = set(from_obj['attributes'].keys())
+        to_keys = set(to_obj['attributes'].keys())
+        removed = from_keys - to_keys
+        added = to_keys - from_keys
+        shared = from_keys & to_keys
+
+        events = []
+        for key in removed:
+          # we don't really have a "remove" event - not sure this ever happens even -
+          # but treat it as equivalent to setting to null
+          #events.append(Document._event_for_attribute_change(from_obj, key, None, value_refs))
+          raise RuntimeError("'RemoveAttribute' events not supported")
+
+        for key in added:
+            new_value = to_obj['attributes'][key]
+            events.append(Document._event_for_attribute_change(from_obj, key, new_value, value_refs, doc))
+
+        for key in shared:
+            old_value = from_obj['attributes'].get(key, None)
+            new_value = to_obj['attributes'].get(key, None)
+
+            if old_value is None and new_value is None:
+                continue
+
+            if old_value is None or new_value is None or old_value != new_value:
+                event = Document._event_for_attribute_change(from_obj, key, new_value, value_refs, doc)
+                events.append(event)
+
+        return events
+
+    # we use this to detect changes during document deserialization
+    # (in model constructors and initializers)
+    @classmethod
+    def _compute_patch_between_json(cls, from_json, to_json, to_doc):
+
+        def refs(json):
+          result = {}
+          for obj in json['roots']['references']:
+            result[obj['id']] = obj
+          return result
+
+        from_references = refs(from_json)
+        from_roots = {}
+        from_root_ids = []
+        for r in from_json['roots']['root_ids']:
+          from_roots[r] = from_references[r]
+          from_root_ids.append(r)
+
+        to_references = refs(to_json)
+        to_roots = {}
+        to_root_ids = []
+        for r in to_json['roots']['root_ids']:
+          to_roots[r] = to_references[r]
+          to_root_ids.append(r)
+
+        from_root_ids.sort()
+        to_root_ids.sort()
+
+        if set(from_root_ids) - set(to_root_ids):
+          # this would arise if someone does add_root/remove_root during
+          # document deserialization, hopefully they won't ever do so.
```

Comment:
Looks like my coffeescript you ported this from is broken and only sees removed roots not added roots, so it silently gets confused with added roots. The conditional both here and in coffee needs to detect any difference in the two root sets. Maybe `if from - to: raise "can't remove roots"` and `if to - from: raise "can't add roots"`